### ISSUE TITLE
replace sbt installer with coursier

### ISSUE THREAD

**Issue by aemoncannon** [ensime/ensime-emacs#269]
Presently we rely on sbt for bootstrapping. Sbt downloads all the dependencies and generates the commandline to start the ensime server. This has a few problems:
- Sbt is slow. Initial ensime boot takes a while, largely because sbt downloads all the deps piecemeal. 
- Sbt is flaky. One of our most common issue tracker responses is 'sbt sucks. delete ivy cache'. Had to give myself this advice last night.
- We want to support Java programmers. Java programmers will be like, 'huh'?

Possible alternative:
We have our build bot push a github tag for each release, including the assembly.jar and init scripts (shell, batch, etc). Github large binary support helps here. Clients will do sparse clones of ensime/ensime-server to grab the assembly/scripts they need.
##

**Comment by fommil** [comment]
I've been thinking similar thoughts, but I can't think of a suitable alternative without creating our own launcher. That's why i created the assembly jars as an alternative. Java developers are more likely to have maven installed, so we could have the maven equivalent of the bootstrapper. In either case, I'm keen the make SBT an optional component and put all references to SBT in its own file.

**Comment by fommil** [comment]
BTW, if we could download the assembly jar from Emacs using a "only download if changed" header, we don't even need git. Although the git idea is also a good one (and means we wouldn't have to use typesafe space). What's the minimal version of git required, and is it stock git?

**Comment by fommil** [comment]
Disadvantage of this is that upgrades force a full download, not just the delta. Arguably, that's still faster than the epic fail that is ivy resolution.

**Comment by fommil** [comment]
bugger, I just remembered that the assembly jars are tied to a very specific version of scala (including minor number). That makes for a **lot** of assembly jars if we want to support everybody this way.

**Comment by aemoncannon** [comment]
womp womp : (

**Comment by fommil** [comment]
a better way? https://github.com/alexarchambault/coursier/issues/69#issuecomment-158758586

**Comment by aemoncannon** [comment]
Would that still rely on ivy?

**Comment by fommil** [comment]
nope, it's a rewrite of that entire layer and **significantly** faster. If typesafe knew a damn thing about what every scala developer on the planet wants, they'd reimplement the entire dependency resolution mechanism in `sbt` to make use of coursier.

**Comment by alexarchambault** [comment]
@fommil @aemoncannon The launcher of coursier is now documented in its [README](https://github.com/alexarchambault/coursier). It seems that you need dependency overrides too, isn't it? (according to `ensime-startup.el`). They're not handled yet by coursier, but it shouldn't be a big deal to add support for it (follow https://github.com/alexarchambault/coursier/issues/17).

**Comment by fommil** [comment]
Sweet!

**Comment by fommil** [comment]
I had a thought which would allow us to use @aemoncannon's original suggestion.

If we build our assembled jars and exclude all the `scala-` jars **and** introduce an extra (optional) field in the `.ensime` file for those jars, then we could ship an assembly jar which is scala minor-version independent!

That would give us three launchers:

1. `sbt` based
2. coursier launcher (if the `jar` is small enough, i.e. <50k, we could distribute with `ensime-emacs`)
3. the assembly jar with scala not included

**Comment by alexarchambault** [comment]
@fommil The coursier launcher is now < 8 kB! I hope that's fine for you! :-)

**Comment by fommil** [comment]
@alexarchambault that's awesome! :-) looking forward to the dependency override support

**Comment by fommil** [comment]
@alexarchambault btw, can the launcher return just the classpath? We don't want to have a JRE between us and the launched app as that consumes real RAM.

**Comment by alexarchambault** [comment]
It can, see the [`classpath` command](https://github.com/alexarchambault/coursier#classpath). Note that as the coursier output is a bit messy for now, you have to keep only the last line of its output (as illustrated in the example in the README), but then it's fine.

**Comment by fommil** [comment]
Hmm, it just occurred to me that we could move the classpath responsibility entirely into the `.ensime` file by adding an extra parameter `:ensime-server-jars`. Hmm. Worth considering. https://github.com/ensime/ensime-sbt/issues/119

**Comment by fommil** [comment]
I've been thinking about this one and there are two disadvantages to shifting responsibility to the build tool:

1. The client is no longer in direct control of updating the server
2. The various build backends need to be updated, not just SBT.


I think we can solve 2 by poking the relative maintainers (we might need to provide a build of SBT for early 0.13.X releases).

We could also workaround 1 by changing ensime-update to run the necessary build command. This could be frustrating for people as gen-ensime can be quite slow.

**Comment by fommil** [comment]
we no longer need to include the scala compiler (version specific) jars in the assembly jar https://github.com/ensime/ensime-sbt/issues/119 which is very interesting... this is the preferred corporate installation mechanism, but could easily replace sbt as our preferred way.

**Comment by fommil** [comment]
just so we're all linked up, we're waiting for https://github.com/alexarchambault/coursier/issues/69 to really minimise the jar size before we'd consider distributing coursier with ensime-emacs

**Comment by aemoncannon** [comment]
ah thanks, was wondering where we stood here. can't wait for this

**Comment by fommil** [comment]
I'm actually using the assembly jars :smile:

**Comment by fommil** [comment]
coursier can't be the default because it doesn't work on java 6 https://github.com/alexarchambault/coursier/issues/137

I don't think we're blocked anymore, we should be able to do this. I'm using the assembly jars at work so this isn't a top priority for me, but could be a good rainy saturday task.

**Comment by fommil** [comment]
@alexarchambault thanks for looking into the java 6 stuff! Is there a standalone bootstrapper for java 6?

**Comment by alexarchambault** [comment]
See the link on the [java-6 branch page](https://github.com/alexarchambault/coursier/tree/topic/java-6) (that I just updated).

In a few days (as soon as I'll have clean and push things that I have locally), the launcher will download the proguarded JAR - and only it, and launch it, instead of its 24 MB of dependencies.

**Comment by fommil** [comment]
damn, I forget to write down the script that I ran at the scalax hack day to download `ensime-server`.

**Comment by alexarchambault** [comment]
I found this somewhere:
```
SCALA_VERSION=${SCALA_VERSION:-2.11.7}

./coursier \
  fetch \
  -r ivy2Cache \
  -r http://bits.netbeans.org/nexus/content/groups/netbeans \
  -r https://oss.sonatype.org/content/repositories/snapshots \
  org.ensime:ensime_2.11:0.9.10-SNAPSHOT \
  -V org.scala-lang:scala-compiler:$SCALA_VERSION \
  -V org.scala-lang:scala-library:$SCALA_VERSION \
  -V org.scala-lang:scala-reflect:$SCALA_VERSION \
  -V org.scala-lang:scalap:$SCALA_VERSION
```

**Comment by fommil** [comment]
whatever the latest URL is

```
curl -L -o coursier https://git.io/v2L2P && chmod +x coursier && ./coursier --help
```

then

```
./coursier fetch org.ensime:ensime_2.11:0.9.10-SNAPSHOT org.scala-lang:scala-compiler:2.11.7 org.scala-lang:scala-reflect:2.11.7 org.scala-lang:scala-library:2.11.7 org.scala-lang:scala-reflect:2.11.7 org.scala-lang:scalap:2.11.7 -r file:///$HOME/.m2/repository -r https://oss.sonatype.org/content/repositories/snapshots -r https://jcenter.bintray.com/ 2> /dev/null
```

which even supports typelevel/typesafe scala installs in maven.

W0000000T!


and to confirm it works with java6

```
/usr/lib/jvm/java-6-oracle/bin/java -jar coursier fetch org.ensime:ensime_2.11:0.9.10-SNAPSHOT org.scala-lang:scala-compiler:2.11.7 org.scala-lang:scala-reflect:2.11.7 org.scala-lang:scala-library:2.11.7 org.scala-lang:scala-reflect:2.11.7 org.scala-lang:scalap:2.11.7 -r file:///$HOME/.m2/repository -r https://oss.sonatype.org/content/repositories/snapshots -r https://jcenter.bintray.com/ 2> /dev/null
```

**Comment by alexarchambault** [comment]
@fommil Shouldn't it support multiple scala versions? Then you should really add `-V` in front of each of the `org.scala-lang` dependencies, to force their (possibly lower) versions.

**Comment by fommil** [comment]
ok, so this is now unblocked and I'm bumping to priority. Things that need to happen

- [ ] allow jars to be added to our melpa config (PR to melpa)
- [ ] @alexarchambault is there a source zip/jar for the launcher jar? would be nice to dist (we're GPL, which escalates all deps to GPL)
- [ ] create a defcustom `ensime-server-classpath-provider` which can be one of `'sbt 'coursier 'assembly` where each references a function having the signature `(ensime-file callback)` where `callback` takes a list of the classpath entries, unordered (it can then monkey prioritise them).
- [ ] refactor the current code to fit the above API (and make coursier the default)
- [ ] ...
- [ ] Profit!

**Comment by fommil** [comment]
equivalents in:

- Atom: https://github.com/ensime/ensime-atom/issues/183
- Vim: https://github.com/ensime/ensime-vim/issues/167
- Sublime: https://github.com/ensime/ensime-sublime/issues/101

**Comment by fommil** [comment]
this is amazeballs

**Comment by alexarchambault** [comment]
@fommil Note that the launcher does *not* automatically update the snapshot metadata/artifacts once it has them in cache. It has to be launched with `-m update-changing` from time to time for that.

**Comment by fommil** [comment]
ok, we want that every time

**Comment by alexarchambault** [comment]
The sources of the bootstrap are simply what's in https://github.com/alexarchambault/coursier/tree/master/bootstrap/src/main/java/coursier. The JAR itself is this put in a JAR, along with the URLs of the dependencies of the launched app (here, coursier).

**Comment by fommil** [comment]
oh, that downloads *everything* every time, ouchies

**Comment by fommil** [comment]
well I guess we could wipe the cache of the snapshots

**Comment by fommil** [comment]
i.e. `rm -rf $HOME/.coursier/cache/v1/https/oss.sonatype.org/content/repositories/snapshots/org/ensime`

**Comment by alexarchambault** [comment]
Yeah, it should work. I remember `-m update-changing` already not doing the right thing some time ago. That should be fixed.

**Comment by fommil** [comment]
an incremental update of the ensime server takes 10 seconds on my machine. **wow**

**Comment by alexarchambault** [comment]
`-m update-changing` should not even try to update all the netbeans / `RELEASE???` things.

**Comment by fommil** [comment]
`update-changing` basically downloaded everything again. I'm actually ok with manually wiping the files, although that will fail on windows if an ensime server is running (but it always will 
no matter what you try to do).

**Comment by alexarchambault** [comment]
Yeah, I created an issue for that (https://github.com/alexarchambault/coursier/issues/154).

**Comment by alexarchambault** [comment]
It doesn't download everything, it only checks (head request) all.

**Comment by fommil** [comment]
why was this not flagged "OMG! Ponies!". We're slipping.

**Comment by jvican** [comment]
is this ready to be used? have you already integrated this with emacs?

**Comment by fommil** [comment]
@jvican we need someone to do it in emacs

**Comment by fommil** [comment]
@bodil pointed out at the Church of Emacs (yes that's a real thing @aemoncannon , wish you lived in London?) that we could use the async library http://melpa.org/#/web as is done for the daily otter https://github.com/bodil/ohai-emacs/blob/02c5cf319a86bcda327bad6d1b4f3c5b2bdcdc35/modules/ohai-splash.el#L61-L65

We could use that to provide a minimal coursier, or a full coursier, or an ensime assembly jar.

**Comment by fommil** [comment]
closed in favour of #560
